drivers/gpio/gpio_name.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2022 Google Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <devicetree.h>
+#include <drivers/gpio/gpio_name.h>
+#include <string.h>
+#include <logging/log.h>
+
+#ifdef CONFIG_GPIO_SHELL
+#include <shell/shell.h>
+#include <stdio.h>
+#endif
+
+/*
+ * Try to determine the base node containing the GPIO nodes.
+ * Mostly the GPIO controllers are child nodes of the
+ * 'soc' node.
+ * A node label 'gpio_controller_parent' may be set
+ * on the parent node.
+ */
+
+#if DT_NODE_EXISTS(DT_ALIAS(gpio_controller_parent))
+
+#define GPIO_BASE_NODE DT_ALIAS(gpio_controller_parent)
+
+#elif DT_NODE_EXISTS(DT_PATH(soc))
+
+#define GPIO_BASE_NODE DT_PATH(soc)
+
+#elif DT_HAS_COMPAT_STATUS_OKAY(zephyr_posix)
+
+#define GPIO_BASE_NODE DT_ROOT
+
+#else
+
+#error "Unable to determine node containing GPIO controllers"
+
+#endif
+
+/*
+ * For this GPIO controller node, create a string array containing the
+ * gpio-line-names strings, and name the array using the node ID.
+ */
+#define NAME_DEFN(id) \
+	static const char *const id[] = DT_PROP(id, gpio_line_names)
+
+/*
+ * For GPIO controller nodes that have the gpio-line-names property,
+ * create the string array for the names.
+ */
+#define NAME_ENTRY(id)					   \
+	COND_CODE_1(DT_NODE_HAS_PROP(id, gpio_line_names), \
+		    (NAME_DEFN(id);),			   \
+		    ())
+
+/*
+ * Walk through all the child nodes of the base node, and
+ * for each node that contains a gpio-line-names property,
+ * generated an array of names from this property.
+ */
+DT_FOREACH_CHILD_STATUS_OKAY(GPIO_BASE_NODE, NAME_ENTRY)
+
+/*
+ * Holds a summary of the GPIO controller data captured from DTS
+ * nodes that have the gpio-line-names property.
+ */
+struct gpio_names {
+	const struct device *port;      /* GPIO controller device */
+	const char *const *names;       /* pointer to array of names */
+	uint8_t name_count;             /* Count of names in array */
```

Comment:
The two fields should be moved to `struct gpio_driver_config` in `include/drivers/gpio.h` file.
```
struct gpio_driver_config {
	gpio_port_pins_t port_pin_mask;
#ifdef CONFIG_GPIO_NAME
	const char *const *names;
	uint8_t name_count;
#endif
};
```
That also implies that the fields need to be initialized by the gpio driver.